src/server/repositories/UrlRepository.ts

```diff
@@ -93,44 +132,67 @@ export class UrlRepository implements UrlRepositoryInterface {
     file?: StorableFile,
   ) => Promise<StorableUrl> = async (originalUrl, changes, file) => {
     const { shortUrl } = originalUrl
-    const url = await Url.scope(['defaultScope', 'getClicks']).findOne({
+    let updateParams: any = { ...changes }
+    const url = await Url.scope([
+      'defaultScope',
+      'getClicks',
+      'getTags',
+    ]).findOne({
       where: { shortUrl },
     })
     if (!url) {
       throw new NotFoundError(
         `url not found in database:\tshortUrl=${shortUrl}`,
       )
     }
-
-    const newUrl: UrlType = await sequelize.transaction(async (t) => {
+    const urlDto = this.urlMapper.persistenceToDto(url)
+    const newUrl = await sequelize.transaction(async (t) => {
+      if (
+        changes.tags &&
+        !_.isEqual(_.sortBy(urlDto.tags), _.sortBy(changes.tags))
+      ) {
+        const newTags = await this.tagRepository.upsertTags(changes.tags, t)
+        // @ts-ignore provided by Sequelize during runtime
+        await url.setTags(newTags, { transaction: t })
+        updateParams = {
+          ...updateParams,
+          tagStrings: updateParams.tags.join(tagSeparator),
+        }
+      }
       if (!url.isFile) {
-        await url.update(changes, { transaction: t })
+        await url.update(updateParams, { transaction: t })
       } else {
         let currentKey = this.fileBucket.getKeyFromLongUrl(url.longUrl)
         if (file) {
           const newKey = file.key
           await url.update(
-            { ...changes, longUrl: this.fileBucket.buildFileLongUrl(newKey) },
+            {
+              ...updateParams,
+              longUrl: this.fileBucket.buildFileLongUrl(newKey),
+            },
             { transaction: t },
           )
           await this.fileBucket.setS3ObjectACL(currentKey, Private)
           await this.fileBucket.uploadFileToS3(file.data, newKey, file.mimetype)
           currentKey = newKey
         } else {
-          await url.update({ ...changes }, { transaction: t })
+          await url.update(updateParams, { transaction: t })
         }
-        if (changes.state) {
+        if (updateParams.state) {
           await this.fileBucket.setS3ObjectACL(
             currentKey,
-            changes.state === StorableUrlState.Active ? Public : Private,
+            updateParams.state === StorableUrlState.Active ? Public : Private,
           )
         }
       }
-      return url
+      // Do a fresh read which eagerly loads the associated tags field.
+      return Url.scope(['defaultScope', 'getClicks', 'getTags']).findOne({
+        where: { shortUrl },
+        transaction: t,
+      })
     })
-
     this.invalidateCache(shortUrl)
-
+    if (!newUrl) throw new Error('Newly-updated url is null')
```

Comment:
how would this come about? sounds like something we should catch earlier?

Reply:
There could be a failed transaction and the cause this to be null.

Reply:
got it - how is this handled downstream? maybe the error should be more informative like "Url update failed, transaction `t` rolled back" or something?

Reply:
I am following the pattern in createURL right now, I don't think it's necessary to show to end user the message like `transaction failed, etc`. We should already have some logs in RDS to tell us what exactly has happened.

Reply:
sure sounds good for now